runtime/include/cloe/databroker.hpp

```diff
@@ -0,0 +1,1254 @@
+/*
+ * Copyright 2023 Robert Bosch GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * \file cloe/databroker.cpp
+ * \see  cloe/databroker.hpp
+ *
+ * Classes:
+ *   DataBroker:
+ *     A central registry for (type erased) signals.
+ *   
+ *   Signal:
+ *     A type-erased abstraction of a signal.
+ *  
+ *   Container<T>
+ *     An optional container storing the value of a signal and abstracting the interaction with Signal.
+ *   
+ * Background:
+ *   Real world simulations utilize manifolds of signals (variables) and coresponding types. 
+ *   Still the very basic problem around a single signal (variable) boils down to CRUD
+ *   - declaring a signal (Create)
+ *   - reading its value / receiving value-changed events (Read)
+ *   - writing its value / triggering value-changed events (Update)
+ *   - NOT NEEDED: undeclaring a signal (Delete)
+ *
+ * Concept:
+ *   Abstract CRU(D) operations by registering signals in a uniform way by their name.
+ */
+
+#pragma once
+#ifndef CLOE_DATABROKER_HPP_
+#define CLOE_DATABROKER_HPP_
+
+#include <any>
+#include <map>
+#include <vector>
+#include <memory>
+#include <functional>
+#include <stdexcept>
+
+#include <boost/preprocessor/cat.hpp>
+#include <boost/preprocessor/list/for_each_i.hpp>
+#include <boost/preprocessor/seq/for_each_product.hpp>
+
+namespace cloe
+{
+
+namespace databroker
+{
+
+template <typename TValue>
+using on_value_change_event_std_function_t = std::function<void(const TValue*)>;
+template <typename TValue>
+using on_value_change_event_function_t = void(*)(const TValue*);
+template <typename TClass, typename TValue>
+using on_value_change_event_member_function_t = void(TClass::*)(const TValue*);
+
+} // namespace databroker
+
+/**
+ * Predicate which determines whether a type is wanted/unwanted as signal datatype
+ */
+template<typename TValue>
+constexpr bool is_wanted_type = std::is_same<TValue, typename std::remove_cv<typename std::remove_pointer<typename std::remove_reference<TValue>::type>::type>::type>::value;
+/**
+ * SFINAE template to determine return-type for enabling functions consuming wanted type T
+ */
+template<typename TValue, typename TReturn>
+using enable_if_is_wanted = std::enable_if<is_wanted_type<TValue>, TReturn >;
+
+#define __HAVE_BETTER_ERROR_MESSAGES 1

```

Comment:
Isn't double underscore reserved?

See: https://stackoverflow.com/questions/228783/what-are-the-rules-about-using-an-underscore-in-a-c-identifier.

Just use `CLOE_IMPROVE_COMPILER_ERRORS` as a name with the ifndef / define / endif at the top, which lets us set it from CMake / Conan.

Reply:
Done.